x/programs/cmd/simulator/src/lib.rs

```diff
@@ -85,23 +85,52 @@ impl Step {
 #[serde(rename_all = "lowercase")]
 #[serde(tag = "type", content = "value")]
 pub enum Key {
+    #[serde(serialize_with = "base64_encode")]
     Ed25519(String),
+    #[serde(serialize_with = "base64_encode")]
     Secp256r1(String),
 }
 
 // TODO:
 // add `Cow` types for borrowing
-#[serde_as]
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
-#[serde(rename_all = "lowercase", tag = "type", content = "value")]
+#[derive(Clone, Debug, PartialEq)]
 pub enum Param {
-    U64(#[serde_as(as = "DisplayFromStr")] u64),
+    U64(u64),
     String(String),
     Id(Id),
-    #[serde(untagged)]
     Key(Key),
 }
 
+#[derive(Serialize)]
+#[serde(rename_all = "lowercase", tag = "type", content = "value")]
+enum StringParam {
+    U64(String),
+    String(String),
+    Id(String),
+}
+
+impl Serialize for Param {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Param::U64(num) => {
+                Serialize::serialize(&StringParam::U64(b64.encode(num.to_le_bytes())), serializer)
+            }
+            Param::String(text) => {
+                Serialize::serialize(&StringParam::String(b64.encode(text)), serializer)
+            }
+            Param::Id(id) => {
+                let num: &usize = id.into();
+                let id = format!("step_{}", num);
+                Serialize::serialize(&StringParam::Id(b64.encode(id)), serializer)
+            }
+            Param::Key(key) => Serialize::serialize(key, serializer),
+        }
+    }
```

Comment:
This is a little more complex than what I had suggested. Also, you really don't need the `step_`. In json, the `"type"` field will be `"id"`, right? that means that `step_` is redundant.